lib/Widgets/Dialog.vala

```diff
@@ -0,0 +1,202 @@
+            action_area.add (widget);
+        }
+
+        var layout = new Gtk.Grid () {
+            margin = 12,
+            row_spacing = 24
+        };
+        layout.attach (content_area, 0, 0);
+        layout.attach (action_area, 0, 1);
+
+        var event_box = new Gtk.EventBox ();
+        event_box.add (layout);
+
+        deletable = false;
+        type_hint = Gdk.WindowTypeHint.DIALOG;
+        window_position = Gtk.WindowPosition.CENTER_ON_PARENT;
+
+        add (event_box);
+
+        event_box.button_press_event.connect ((event) => {
+        if (event.button == Gdk.BUTTON_PRIMARY) {
+            begin_move_drag ((int) event.button, (int) event.x_root, (int) event.y_root, event.time);
+                return Gdk.EVENT_STOP;
+            }
+
+            return Gdk.EVENT_PROPAGATE;
+        });
+    }
+
+    public override bool key_press_event (Gdk.EventKey event) {
+        if (event.keyval == Gdk.Key.Escape) {
+            response (Gtk.ResponseType.CLOSE);
+            destroy ();
+        }
+
+        return base.key_press_event (event);
+    }
+
+    /**
+    * Adds a button with the given text and sets things up so that clicking the button will emit the response signal with the given response_id.
+    */
+    public Gtk.Widget add_button (string button_text, int response_id) {
+        var button = new Gtk.Button.with_label (button_text) {
+            use_underline = true
+        };
+
+        return add_action_widget (button, response_id);
+    }
+
+    /**
+    * Adds an activatable widget to the action area of {@link Granite.Dialog}, connecting a signal handler that will emit the {@link Granite.Dialog.response} signal on the dialog when the widget is activated.
+    */
+    public Gtk.Widget add_action_widget (Gtk.Widget widget, int response_id) {
+        widget.button_release_event.connect (() => {
+            response (response_id);
+
+            return Gdk.EVENT_STOP;
+        });
+
+        // It's possible that this is called before action_area is constructed such as in Granite.MessageDialog
+        if (action_area == null) {
+            if (action_widgets == null) {
+                action_widgets = new List<Gtk.Widget> ();
+            }
+
+            action_widgets.append (widget);
+        } else {
+            action_area.add (widget);
+        }
+
+        return widget;
+    }
+
+    /**
+    * Blocks in a recursive main loop until {@link Granite.Dialog} either emits the {@link Granite.Dialog.response} signal, or is destroyed.
+    * If the dialog is destroyed during the call to run, run returns {@link Gtk.ResponseType.NONE}. Otherwise, it returns the response ID from the {@link Granite.Dialog.response} signal emission.
+    * Before entering the recursive main loop, run calls {@link Gtk.Widget.show_all} on the dialog for you.
+    * After run returns, you are responsible for hiding or destroying the dialog if you wish to do so.
+    */
+    public int run () {
```

Comment:
It sounds like this is gone in Gtk4, so we should mark it as deprecated right away to help with the transition

Reply:
Agree :+1: